### PR TITLE
Hacky fix to make mac wheels 

### PR DESCRIPTION
**Pull request by DMRobertson**
A desperate attempt to make CI work again after #14949.

Tested in #15015.

My hypothesis is:

- poetry-core 1.3.2 builds a wheel that says it's for `macosx_10_16` under the CI env
- poetry-core 1.5.0 (maybe 1.4.0?) builds a wheel that says it's for `macosx_11_0` under the CI
- the version of pip in the CI env considers that not installable

### CODE REVIEW

File: .github/workflows/release-artifacts.yml [hunk 1]
```diff
@@ -148,7 +148,7 @@ jobs:
         env:
           # Skip testing for platforms which various libraries don't have wheels
           # for, and so need extra build deps.
-          CIBW_TEST_SKIP: pp3{7,9}-* *i686* *musl*
+          CIBW_TEST_SKIP: pp3*-* *i686* *musl*
```
Comment (clokep) [inline .github/workflows/release-artifacts.yml:151]: Why this change?

File: changelog.d/15019.misc [hunk 2]
```diff
@@ -0,0 +1 @@
+Workaround packaging pain when making OSX wheels.
```
Comment (clokep) [inline changelog.d/15019.misc:1]:
```suggestion
Fix creation of wheels on macOS.
```

File: .ci/scripts/auditwheel_wrapper.py [hunk 3]
```diff
@@ -50,7 +50,16 @@ def cpython(wheel_file: str, name: str, version: Version, tag: Tag) -> str:
...
+    # HACK: it seems that some older versions of pip will consider a wheel marked
+    # as macosx_11_0 as incompatible with Big Sur. I haven't done the full archaeology
+    # here; there are some clues in
+    #     https://github.com/pantsbuild/pants/pull/12857
+    #     https://github.com/pypa/pip/issues/9138
+    #     https://github.com/pypa/packaging/pull/319
+    # Empirically this seems to work.
+    # I hate Python packaging, and you should too. Guido, why have you forsaken me?
+    platform = tag.platform.replace("macosx_11_0", "macosx_10_16")
+    abi3_tag = Tag(tag.interpreter, "abi3", platform)
```
Comment (clokep) [inline .ci/scripts/auditwheel_wrapper.py:62]:
```suggestion
    # HACK: it seems that some older versions of pip will consider a wheel marked
    # as macosx_11_0 as incompatible with Big Sur. I haven't done the full archaeology
    # here; there are some clues in
    #     https://github.com/pantsbuild/pants/pull/12857
    #     https://github.com/pypa/pip/issues/9138
    #     https://github.com/pypa/packaging/pull/319
    # Empirically this seems to work, note that macOS 11 and 10.16 are the same,
    # both versions are valid for backwards compatibility.
    platform = tag.platform.replace("macosx_11_0", "macosx_10_16")
    abi3_tag = Tag(tag.interpreter, "abi3", platform)
```

File: .github/workflows/release-artifacts.yml (hunk 1 above)
Comment (DMRobertson) [reply .github/workflows/release-artifacts.yml:151]: Per the commit message, I saw a CI failure when testing this out where the linux pypy3.8 build failed: https://github.com/matrix-org/synapse/actions/runs/4117573006/jobs/7109038672

I don't see why pypy3.8 gets special treatment here, but I could also just drop this rather than try to sneak it in.

Comment (clokep) [reply .github/workflows/release-artifacts.yml:151]: I failed to review commit by commit.

**Review by clokep — APPROVED** [review]

File: .github/workflows/release-artifacts.yml (hunk 1 above)
Comment (DMRobertson) [reply .github/workflows/release-artifacts.yml:151]: Equally, I failed to point it out!

**Comment by DMRobertson** [comment]
Cross-refs to things I read while investigating this:

- https://github.com/pypa/cibuildwheel/issues/1410
- https://github.com/pantsbuild/pants/pull/12857
- https://github.com/pypa/pip/issues/9138
- https://github.com/pypa/packaging/pull/319

File: .github/workflows/release-artifacts.yml (hunk 1 above)
Comment (erikjohnston) [reply .github/workflows/release-artifacts.yml:151]: > I don't see why pypy3.8 gets special treatment here, but I could also just drop this rather than try to sneak it in.

I think the idea was to test *a* pypy version, mostly.

Comment (DMRobertson) [reply .github/workflows/release-artifacts.yml:151]: Having slept on it, I wonder if that failure was related to the recent cryptography release. Maybe there is no longer a wheel available for that platform?

We built

```
 Fixed-up wheel written to /tmp/cibuildwheel/repaired_wheel/matrix_synapse-1.77.0rc1-pp38-pypy38_pp73-manylinux_2_17_x86_64.manylinux2014_x86_64.whl
```

Which presumably relied on there being a cryptography wheel for that system. But the latest release of cryptography has no such wheel:

```
$ curl -s https://pypi.org/project/cryptography/39.0.1/#files | grep -o "cryptography.*-pp38.*\.whl"
cryptography-39.0.1-pp38-pypy38_pp73-win_amd64.whl
cryptography-39.0.1-pp38-pypy38_pp73-win_amd64.whl
cryptography-39.0.1-pp38-pypy38_pp73-macosx_10_12_x86_64.whl
cryptography-39.0.1-pp38-pypy38_pp73-macosx_10_12_x86_64.whl
cryptography-39.0.1-pp38-pypy38_pp73-win_amd64.whl
cryptography-39.0.1-pp38-pypy38_pp73-win_amd64.whl
cryptography-39.0.1-pp38-pypy38_pp73-macosx_10_12_x86_64.whl
cryptography-39.0.1-pp38-pypy38_pp73-macosx_10_12_x86_64.whl

$ curl -s https://pypi.org/project/cryptography/39.0.1/#files | grep -o "cryptography.*-pp38.*linux.*\.whl"
```

Not sure why. https://github.com/pyca/cryptography/pull/7795 stopped building PyPy3.7 wheels, and I can't see any recent changes in Cryptography's CI.

There do seem to be pypy 3.9 wheels though:

```
curl -s https://pypi.org/project/cryptography/39.0.1/#files | grep -o "cryptography.*-pp39.*linux.*\.whl"
cryptography-39.0.1-pp39-pypy39_pp73-manylinux_2_28_x86_64.whl
cryptography-39.0.1-pp39-pypy39_pp73-manylinux_2_28_x86_64.whl
cryptography-39.0.1-pp39-pypy39_pp73-manylinux_2_28_aarch64.whl
cryptography-39.0.1-pp39-pypy39_pp73-manylinux_2_28_aarch64.whl
cryptography-39.0.1-pp39-pypy39_pp73-manylinux_2_24_x86_64.whl
cryptography-39.0.1-pp39-pypy39_pp73-manylinux_2_24_x86_64.whl
cryptography-39.0.1-pp39-pypy39_pp73-manylinux_2_17_aarch64.manylinux2014_aarch64.manylinux_2_24_aarch64.whl
cryptography-39.0.1-pp39-pypy39_pp73-manylinux_2_17_aarch64.manylinux2014_aarch64.manylinux_2_24_aarch64.whl
cryptography-39.0.1-pp39-pypy39_pp73-manylinux_2_28_x86_64.whl
cryptography-39.0.1-pp39-pypy39_pp73-manylinux_2_28_x86_64.whl
cryptography-39.0.1-pp39-pypy39_pp73-manylinux_2_28_aarch64.whl
cryptography-39.0.1-pp39-pypy39_pp73-manylinux_2_28_aarch64.whl
cryptography-39.0.1-pp39-pypy39_pp73-manylinux_2_24_x86_64.whl
cryptography-39.0.1-pp39-pypy39_pp73-manylinux_2_24_x86_64.whl
cryptography-39.0.1-pp39-pypy39_pp73-manylinux_2_17_aarch64.manylinux2014_aarch64.manylinux_2_24_aarch64.whl
cryptography-39.0.1-pp39-pypy39_pp73-manylinux_2_17_aarch64.manylinux2014_aarch64.manylinux_2_24_aarch64.whl
```

So I'll try testing the PyPy 3.9 wheel in CI.
